packages/protocol/contracts/bridge/libs/LibBridgeRecall.sol

```diff
@@ -0,0 +1,160 @@
+        LibBridgeData.State storage state,
+        AddressResolver resolver,
+        IBridge.Message calldata message,
+        bytes calldata proof
+    )
+        internal
+    {
+        if (message.owner == address(0)) {
+            revert B_OWNER_IS_NULL();
+        }
+
+        if (message.srcChainId != block.chainid) {
+            revert B_WRONG_CHAIN_ID();
+        }
+
+        bytes32 msgHash = message.hashMessage();
+
+        if (
+            !LibBridgeStatus.isMessageFailed(
+                resolver, msgHash, message.destChainId, proof
+            )
+        ) {
+            revert B_MSG_NOT_FAILED();
+        }
+
+        if (
+            state.recallStatus[msgHash]
+                == LibBridgeData.RecallStatus.FULLY_RECALLED
+        ) {
+            // Both ether and tokens are released
+            revert B_ETHER_RELEASED_ALREADY();
+        }
+
+        uint256 releaseAmount;
+
+        if (
+            state.recallStatus[msgHash]
+                == LibBridgeData.RecallStatus.NOT_RECALLED
+        ) {
+            // Release ETH first
+            state.recallStatus[msgHash] =
+                LibBridgeData.RecallStatus.ETH_RELEASED;
+
+            releaseAmount = message.depositValue + message.callValue;
+
+            if (releaseAmount > 0) {
+                address ethVault = resolver.resolve("ether_vault", true);
+                // if on Taiko
+                if (ethVault != address(0)) {
+                    EtherVault(payable(ethVault)).releaseEther(
+                        message.owner, releaseAmount
+                    );
+                } else {
+                    // if on Ethereum
+                    (bool success,) =
+                        message.owner.call{ value: releaseAmount }("");
+                    if (!success) {
+                        revert B_FAILED_TRANSFER();
+                    }
+                }
+            }
+        }
+        //2nd stage is releasing the tokens
+        if (
+            state.recallStatus[msgHash]
+                == LibBridgeData.RecallStatus.ETH_RELEASED
+        ) {
+            if (message.sender.isContract()) {
+                if (isRecallableMessageSender(message.sender)) {
+                    try IRecallableMessageSender((message.sender))
+                        .onMessageRecalled(message) {
+                        state.recallStatus[msgHash] =
+                            LibBridgeData.RecallStatus.FULLY_RECALLED;
+                    } catch { }
+                }
+            } else {
+                state.recallStatus[msgHash] =
+                    LibBridgeData.RecallStatus.FULLY_RECALLED;
+            }
+        }
+        emit MessageRecalled(
+            msgHash, message.owner, releaseAmount, state.recallStatus[msgHash]
+        );
+    }
+
+    function isRecallableMessageSender(address addr)
+        private
+        view
+        returns (bool retVal)
+    {
+        try IERC165Upgradeable(addr).supportsInterface(
```

Comment:
why call the same function twice? search for "try-catch" in the code to see how you can try once and get the return code.